test/integration/header_integration_test.cc

```diff
@@ -643,28 +683,132 @@ TEST_P(HeaderIntegrationTest, TestRouteConfigVirtualHostAndRouteReplaceHeaderMan
+          {"x-route-request", "downstream"},
+          {"x-weighted-cluster-request", "downstream"},
+      },
+      Http::TestHeaderMapImpl{
+          {":authority", "vhost-headers.com"},
+          {"x-routeconfig-request", "downstream"},
+          {"x-vhost-request", "downstream"},
+          {"x-route-request", "downstream"},
+          {"x-weighted-cluster-request", "downstream"},
+          {"x-weighted-cluster-request", "weighted-cluster-1"},
+          {"x-route-request", "route"},
+          {"x-vhost-request", "vhost"},
+          {"x-routeconfig-request", "routeconfig"},
+          {":path", "/vhost-route-and-weighted-clusters"},
+          {":method", "GET"},
+      },
+      Http::TestHeaderMapImpl{
+          {"server", "envoy"},
+          {"content-length", "0"},
+          {":status", "200"},
+          {"x-routeconfig-response", "upstream"},
+          {"x-routeconfig-response-remove", "upstream"},
+          {"x-vhost-response", "upstream"},
+          {"x-vhost-response-remove", "upstream"},
+          {"x-route-response", "upstream"},
+          {"x-route-response-remove", "upstream"},
+          {"x-weighted-cluster-response", "upstream"},
+          {"x-weighted-cluster-response-remove", "upstream"},
+      },
+      Http::TestHeaderMapImpl{
+          {"server", "envoy"},
+          {"x-routeconfig-response", "upstream"},
+          {"x-vhost-response", "upstream"},
+          {"x-route-response", "upstream"},
+          {"x-weighted-cluster-response", "upstream"},
+          {"x-weighted-cluster-response", "weighted-cluster-1"},
+          {"x-route-response", "route"},
+          {"x-vhost-response", "vhost"},
+          {"x-routeconfig-response", "routeconfig"},
+          {":status", "200"},
+      });
+}
+
+// Validates the relationship between route configuration, virtual host, route and weighted cluster
+// header manipulations when replacing.
+TEST_P(HeaderIntegrationTest, TestRouteConfigVirtualHostRouteAndClusterReplaceHeaderManipulation) {
+  initializeFilter(HeaderMode::Replace, true);
+  performRequest(
+      Http::TestHeaderMapImpl{
+          {":method", "GET"},
+          {":path", "/vhost-route-and-weighted-clusters"},
+          {":scheme", "http"},
+          {":authority", "vhost-headers.com"},
+          {"x-routeconfig-request", "downstream"},
+          {"x-vhost-request", "downstream"},
+          {"x-route-request", "downstream"},
+          {"x-weighted-cluster-request", "downstream"},
+          {"x-unmodified", "request"},
+      },
+      Http::TestHeaderMapImpl{
+          {":authority", "vhost-headers.com"},
+          {"x-unmodified", "request"},
+          {"x-weighted-cluster-request", "weighted-cluster-1"},
+          {"x-route-request", "route"},
+          {"x-vhost-request", "vhost"},
+          {"x-routeconfig-request", "routeconfig"},
+          {":path", "/vhost-route-and-weighted-clusters"},
+          {":method", "GET"},
+      },
+      Http::TestHeaderMapImpl{
```

Comment:
I might be missing something in how these tests are run, but why not have the `x-weighted-cluster-response-remove` header here?

Reply:
It's in the previous test. You can think of these tests as existing in append/replace pairs, and I decided not to repeat the testing of header removal in the replace tests.

Reply:
Yep, I was wondering if having the remove behavior tested in both places would ensure that remove behavior works correctly regardless of adding scheme (append/replace). But if you don't think that gives us anything I am good with it.